### ISSUE TITLE
Non-media file attachments/document attachments

### ISSUE THREAD

**Issue by upintheairsheep** [mastodon/mastodon#26420]
### Pitch

Please add functionality for posting a file that cannot be previewed: images, videos, and audio. Twitter users frequently use services like Discord's CDN for posting non-media files.

### Motivation

I personally would benefit from being able to share any file I wish. Additionally, it is important for programmers to share code files to discuss in a corporate setting, and for artists and other creative fields to share files on emerging formats not yet supported or impractical to support.

**Comment by JustDalek** [comment]
I'd rather use google drive for this and post a link

**Comment by upintheairsheep** [comment]
> I'd rather use google drive for this and post a link

Google Drive is not as reliable

**Comment by brendanjones** [comment]
Yes to this. I've a few times recently wanted to share text, excel and pdf files. Basically to share content that doesn't fit in posts due to its length, and/or that contains formatting and content that's not supported by Mastodon posts, and/or doesn't already exist on the internet at an accessible URL. 

One question I'd ask is if there need to be any limitations on this, like file size limitations? It's not like Mastodon is designed to be a file sharing and hosting service (nor should it be; I see this capability as a nice addition, not core functionality), so I imagine allowing it to be one may add to hosting requirements?

Maybe any limits could be something configurable by server admins? This would then be a service-differentiator between servers, just like post character limit is now. 

> I'd rather use google drive for this and post a link

That does not solve the problem for anyone not using Google Drive or any other cloud hosting service.

**Comment by BakaDalek** [comment]
> That does not solve the problem for anyone not using Google Drive or any other cloud hosting service.

My main concern is Mastodon is social networking and not cloud storage. As an end user this sounds awesome, as a host this sounds like a nightmare. 
I have friends that make discord servers with the sole purpose of using them for free storage and file sharing. 

Given discord is a big company nobody bats an eye at this, who cares right?

But give this functionality to decentralized platforms like mastodon, servers with some tighter storage constraints and budgets and you will have end users eating up more storage posting documents for file sharing purposes and not clean up old files. 

For mastodon storage is my #1 concern. I feel like if we took a page from misskey and introduced storage upload limitations per-user accounts (and optionally an integration or native way to reward supporters with some extra storage) I'd be more into loosening up what all can be uploaded.

**Comment by kevATin** [comment]
To me Mastodon is primarily a communication platform, secondarily social media in the traditional sense.
I agree that Mastodon shouldn't be used as cloud storage, but it does already allow uploading lots of files and using up storage space. If I were to host a public instance, I'd probably try to disable images, audio, and video uploads, at least with Mastodon's current feature set.

I don't know how misskey handles it, but i'm very much in favor of optional file upload limits per user.
It should probably be a weekly quota or similar and not a total limit, otherwise users would be forced to delete old posts.

But in turn users (and admins) should be given more freedom on how to use that storage.
I would like to see apps get distributed via posts (.apk, .appimage, .exe, .dmg, etc) and .iCal and .vCard and .eml and compressed archives with multiple files inside.
I'd like to see an upload option that allows any file without modifying it during the upload. Currently you cannot share .png images with arbitrary data inside without it getting broken, so you cannot share Spore game content/saves for example despite them being images.

I would also be very much in favor of the ability to embed custom apps like twitter did in the past, though that feature got largely removed when the entire MS-DOS games library became playable for free. 
This game dev has managed to put their game inside of a tweet and it still works: https://twitter.com/ClayLoam/status/1337488664243277824 
Personally i see this feature connected to attachments, because in an ideal world the game doesn't even link to an external website but is a .html file with javascript inside.


**TLDR**: 
I would like to get a separate "Binary Upload" option for posts, to upload any file without Mastodon modifying it. 
In turn optional upload limits should be introduced per user.

**Comment by upintheairsheep** [comment]
Pretty sure it got removed from twitter for security concerns, someone
could hijack the browser to steal cookies.

**Comment by kevATin** [comment]
> Pretty sure it got removed from twitter for security concerns, someone could hijack the browser to steal cookies.

The feature was available for a while and got shut down right after the MS-DOS thing happened, i think their reasons were pretty obvious at the time.
Besides the example i linked still works on twitter currently.

**Comment by henningninneh** [comment]
I often share screenshots of calls for applications (for conferences, events, jobs etc.) I have to convert the pdf to png (sometimes multiple png files, if there are multiple pages), copy the text, and add it to the image description. Often the text doesn't fit the image description. PDFs are great for accessibility. If you care about people who rely on screen readers, please support PDF upload.

**Comment by upintheairsheep** [comment]
> To me Mastodon is primarily a communication platform, secondarily social media in the traditional sense. I agree that Mastodon shouldn't be used as cloud storage, but it does already allow uploading lots of files and using up storage space. If I were to host a public instance, I'd probably try to disable images, audio, and video uploads, at least with Mastodon's current feature set.
> 
> I don't know how misskey handles it, but i'm very much in favor of optional file upload limits per user. It should probably be a weekly quota or similar and not a total limit, otherwise users would be forced to delete old posts.
> 
> But in turn users (and admins) should be given more freedom on how to use that storage. I would like to see apps get distributed via posts (.apk, .appimage, .exe, .dmg, etc) and .iCal and .vCard and .eml and compressed archives with multiple files inside. I'd like to see an upload option that allows any file without modifying it during the upload. Currently you cannot share .png images with arbitrary data inside without it getting broken, so you cannot share Spore game content/saves for example despite them being images.
> 
> I would also be very much in favor of the ability to embed custom apps like twitter did in the past, though that feature got largely removed when the entire MS-DOS games library became playable for free. This game dev has managed to put their game inside of a tweet and it still works: https://twitter.com/ClayLoam/status/1337488664243277824 Personally i see this feature connected to attachments, because in an ideal world the game doesn't even link to an external website but is a .html file with javascript inside.
> 
> **TLDR**: I would like to get a separate "Binary Upload" option for posts, to upload any file without Mastodon modifying it. In turn optional upload limits should be introduced per user.

Mastodon can share images and videos, and the ability to upload regular files should be up to the server operator, as well as the size limit.

**Comment by exaexa** [comment]
Hi all,

I thought I'd just add a few pretty good use-cases for this:

- attaching a git patch when talking about code
- attaching an .asc (or such) with GPG signature (or other signature) to provide cryptographically verifiable toot source
- similarly, attaching GPG keys
- broadcasting a .torrent file
- sharing geographic information (KML?) or trails

OTOH I certainly hope people wouldn't use this to share big binary blobs (especially not .docx or .exe etc 😅 ) -- almost all sensible (imo) uses of the feature should fit into a few dozen kilobytes....so I guess a separate size limit configuration should be available for this. Preview (or generally any advanced visibility in the browser views) is _not_ required for this to be useful.

Also please note that this is a somewhat tangential approach to the toot size limit discussion that is happening elsewhere:
- most of the usecases listed here could easily fit into a large toot (possibly with a bit of base64 or such). This feature could provide some convenience and UX for the cases that do not look very well timelines (base64 again :) ).
- attaching sensibly-sized markdown (with some minimal rendering and/or editing in the frontend, even lousy) would work around a large portion of the reasons to increase the toot size

Thanks!
-mk
